components/builder-depot/src/backend/s3.rs

```diff
@@ -0,0 +1,189 @@
+use std::io::{BufReader, Read};
+use std::path::{PathBuf, Path};
+use std::str::FromStr;
+
+use hab_core::package::{PackageTarget, Identifiable, PackageIdent};
+use iron::typemap::Key;
+
+use rusoto::{Region, credential::StaticProvider, reactor::RequestDispatcher} ;
+use rusoto_s3::{S3, S3Client, GetObjectRequest, PutObjectRequest, ListObjectsRequest};
+use futures::{Future, Stream};
+
+use DepotUtil;
+use config::S3Cfg;
+
+use error::{Result, Error};
+
+pub struct S3Handler {
+    client: S3Client<StaticProvider, RequestDispatcher>,
+    bucket: String,
+}
+
+impl S3Handler {
+    // The S3 Handler struct contains all of the credential
+    // and target information that we should need to perfom
+    // any backend operations
+    pub fn new(config: S3Cfg) -> Self {
+        let region =
+            match config.endpoint {
+                Some(ref value) => {
+                    Region::Custom { name: "minio_s3".to_owned(), endpoint: value.to_string() }
+                },
+                None => Region::from_str(config.aws_region.as_str()).unwrap()
+        };
+        let aws_id = config.aws_key_id.trim().to_owned();
+        let aws_secret = config.aws_secret_key.trim().to_owned();
+        let cred_provider = StaticProvider::new_minimal(aws_id, aws_secret);
+        let client = S3Client::new(RequestDispatcher::default(), cred_provider, region);
+        let bucket = config.s3bucket_name;
+
+        S3Handler {
+            client: client,
+            bucket: bucket.to_string(),
+        }
+    }
+
+    // Helper function for programmatic creation of
+    // the s3 object key
+    fn s3_key<T>(&self, ident: &T, target: &PackageTarget) -> PathBuf
+    where
+        T: Identifiable,
+    {
+        Path::new(ident.origin())
+            .join(format!("{}", ident.name()))
+            .join(format!("{}", ident.version().unwrap()))
+            .join(format!("{}", ident.release().unwrap()))
+            .join(format!("{}", target.architecture))
+            .join(format!("{}", target.platform))
+            .join(format!("{}-{}-{}-{}-{}-{}.hart", ident.origin(), ident.name(), ident.version().unwrap(), ident.release().unwrap(), target.architecture, target.platform))
+    }
+
+    // Helper function to validate whether or not a package
+    // exists in the configured bucket
+    pub fn exists<T>(&self, ident: &T, target: &PackageTarget) -> Result<bool>
+    where
+        T: Identifiable,
+    {
+        let mut request = ListObjectsRequest::default();
+        let key = self.s3_key(ident, target).to_string_lossy().into_owned();
+        request.bucket = self.bucket.to_owned();
+        match self.client.list_buckets().sync() {
+            Ok(bucket_list) => {
+                match bucket_list.buckets {
+                    Some(bkts) => {
+                        for bucket in bkts {
+                            if bucket.name.unwrap() == request.bucket {
+                                match self.client.list_objects(&request).sync() {
```

Comment:
Some concerns here: 1) synchronous blocking operation;  2) large result set (we are retrieving list of all objects in our S3 bucket?); 3) according to the API, list_objects returns up to a max of 1000.  This doesn't seem feasible if the idea is to iterate through all objects we have stored looking for a key match..

Reply:
Yes, the API list objects limit is definitely going to be a concern. The operations themselves are futures, so these could be performed in an async way. Will have to think about how to accomplish that.

Reply:
Well, maybe we need to think about whether we need/should have an existence check at all.  We essentially are already keeping a secondary index via the metadata in originsrv.  So we could assume that we have the package in S3 already if the originsrv metadata says we do (post-migration).  So this is only a sanity validation for upload that isn't really serving much need in the S3 world.

Reply:
You might be correct about that. It's worth discussing. I think initially i put this validation in here because its the operation we were doing with locally stored artifacts. I guess my question would be, what about a theoretical edge case where the data exists in orignsrv but the package never actually successfully uploaded to s3?

Reply:
If we upload the package first, and only write to originsrv on a successful upload that should not happen.  We just need to be able to idempotently upload to S3.

Reply:
EoD note on this, I haven't touched this bit yet. Planning to fix this in the morning and theoretically remove this whole function.